File: netlify/functions/generate-voucher/generate-voucher.py
import json
import base64
import os
from io import BytesIO

def handler(event, context):
    if event.get('httpMethod') != 'POST':
        return {'statusCode': 405, 'body': 'Method Not Allowed'}

    try:
        body = json.loads(event.get('body', '{}'))
        voucher_type = int(body.get('voucherType', 1))
        code         = str(body.get('code', '')).strip()
        expiry       = str(body.get('expiry', '')).strip()

        if not code or not expiry:
            return {'statusCode': 400, 'body': json.dumps({'error': 'Missing code or expiry'})}

        # Base JPGs live in the vouchers/ folder at the repo root
        base_map = {
            1: '/var/task/vouchers/base_kowhai.jpg',
            2: '/var/task/vouchers/base_ultimate.jpg',
            3: '/var/task/vouchers/base_pohutukawa.jpg',
        }

        base_path = base_map.get(voucher_type)
        if not base_path or not os.path.exists(base_path):
            return {'statusCode': 400, 'body': json.dumps({'error': f'Base image not found for type {voucher_type}'})}

        img_bytes = generate_voucher_image(base_path, code, expiry)

        return {
            'statusCode': 200,
            'headers': {'Content-Type': 'application/json'},
            'body': json.dumps({
                'image': base64.b64encode(img_bytes).decode('utf-8'),
                'mimeType': 'image/jpeg',
            }),
        }

    except Exception as e:
        print(f'Error: {e}')
        return {'statusCode': 500, 'body': json.dumps({'error': str(e)})}


def generate_voucher_image(base_path, code, expiry):
    from PIL import Image, ImageDraw, ImageFont

    img = Image.open(base_path).convert('RGB')
    draw = ImageDraw.Draw(img)

    font_dir    = os.path.dirname(__file__)
    font_path   = os.path.join(font_dir, 'PlayfairDisplay.ttf')
    font_code   = ImageFont.truetype(font_path, 90)
    font_expiry = ImageFont.truetype(font_path, 76)

    draw.text((2280, 3230), code,   font=font_code,   fill=(0, 0, 0))
    draw.text((2321, 3415), expiry, font=font_expiry, fill=(0, 0, 0))

    buf = BytesIO()
    img.save(buf, format='JPEG', quality=95)
    return buf.getvalue()

File: netlify/functions/generate-voucher/generate-voucher.js
const sharp = require('sharp');

exports.handler = async function(event) {
  if (event.httpMethod !== 'POST') {
    return { statusCode: 405, body: 'Method Not Allowed' };
  }

  try {
    const body = JSON.parse(event.body || '{}');
    const voucherType = parseInt(body.voucherType || 1);
    const code        = String(body.code   || '').trim();
    const expiry      = String(body.expiry || '').trim();

    if (!code || !expiry) {
      return { statusCode: 400, body: JSON.stringify({ error: 'Missing code or expiry' }) };
    }

    const baseMap = {
      1: 'https://raw.githubusercontent.com/WildCanvas/wildcanvas/main/netlify/functions/generate-voucher/base_kowhai.jpg',
      2: 'https://raw.githubusercontent.com/WildCanvas/wildcanvas/main/netlify/functions/generate-voucher/base_ultimate.jpg',
      3: 'https://raw.githubusercontent.com/WildCanvas/wildcanvas/main/netlify/functions/generate-voucher/base_pohutukawa.jpg',
    };

    const imageUrl = baseMap[voucherType];
    if (!imageUrl) {
      return { statusCode: 400, body: JSON.stringify({ error: 'Invalid voucher type' }) };
    }

    const imageResponse = await fetch(imageUrl);
    if (!imageResponse.ok) {
      return { statusCode: 500, body: JSON.stringify({ error: 'Could not fetch base image: ' + imageResponse.status }) };
    }
    const imageBuffer = Buffer.from(await imageResponse.arrayBuffer());

    // Resize to 1200px wide first
    const resizedBuffer = await sharp(imageBuffer)
      .resize(1200)
      .jpeg({ quality: 95 })
      .toBuffer();

    // Use DejaVu Sans — guaranteed available on Linux/Netlify
    const svgOverlay = `<svg width="1200" height="1289" xmlns="http://www.w3.org/2000/svg">
      <text x="759" y="1106" font-family="DejaVu Sans, Liberation Sans, Arial, sans-serif" font-size="30" font-weight="bold" fill="black">${escapeXml(code)}</text>
      <text x="773" y="1162" font-family="DejaVu Sans, Liberation Sans, Arial, sans-serif" font-size="25" fill="black">${escapeXml(expiry)}</text>
    </svg>`;

    const outputBuffer = await sharp(resizedBuffer)
      .composite([{ input: Buffer.from(svgOverlay), top: 0, left: 0 }])
      .jpeg({ quality: 95 })
      .toBuffer();

    return {
      statusCode: 200,
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        image:    outputBuffer.toString('base64'),
        mimeType: 'image/jpeg',
      }),
    };

  } catch (err) {
    console.error('generate-voucher error:', err.message);
    return { statusCode: 500, body: JSON.stringify({ error: err.message }) };
  }
};

function escapeXml(str) {
  return String(str)
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&apos;');
}
